ygen/enumgen.go

```diff
@@ -229,47 +238,41 @@ func (s *enumSet) identityBaseKey(i *yang.Identity) string {
 // enumeration). It also returns the type name component of the identification
 // key for use in the name generation, as well as the enumeration type object,
 // if needed.
-func (s *enumSet) enumeratedTypedefKey(e *yang.Entry, noUnderscores, useDefiningModuleForTypedefEnumNames bool) (string, string, *yang.YangType, error) {
-	enumType := e.Type
-	typeName := e.Type.Name
+func (s *enumSet) enumeratedTypedefKey(args resolveTypeArgs, noUnderscores, useDefiningModuleForTypedefEnumNames bool) (string, string, error) {
+	typeName := args.yangType.Name
+	subsumingType := args.subsumingType
+	if subsumingType == nil {
+		// If subsumingType is not provided, then we assume that the given type is not a union subtype.
+		// Then by definition, its subsuming type is the type itself.
+		subsumingType = args.yangType
+	}
 
 	// Handle the case whereby we have been handed an enumeration that is within a
 	// union. We need to synthesise the name of the type here such that it is based on
 	// type name, plus the fact that it is an enumeration.
-	if e.Type.Kind == yang.Yunion {
-		enumTypes := util.EnumeratedUnionTypes(e.Type.Type)
-
-		switch len(enumTypes) {
-		case 1:
-			// Under a union, the correct enumType reference is the enumeration subtype.
-			enumType = enumTypes[0]
-			// We specifically say that this is an enumeration within the leaf.
-			if noUnderscores {
-				typeName = fmt.Sprintf("%sEnum", enumTypes[0].Name)
-			} else {
-				typeName = fmt.Sprintf("%s_Enum", enumTypes[0].Name)
-			}
-		case 0:
-			return "", "", nil, fmt.Errorf("enumerated type had an empty union within it, path: %v, type: %v, enumerated: %v", e.Path(), e.Type, enumTypes)
-		default:
-			return "", "", nil, fmt.Errorf("multiple enumerated types within a single enumeration not supported, path: %v, type: %v, enumerated: %v", e.Path(), e.Type, enumTypes)
+	if subsumingType.Kind == yang.Yunion {
+		// We specifically say that this is an enumeration within the leaf.
+		if noUnderscores {
+			typeName = fmt.Sprintf("%sEnum", subsumingType.Name)
+		} else {
+			typeName = fmt.Sprintf("%s_Enum", subsumingType.Name)
 		}
 	}
-	if e.Node == nil {
-		return "", "", nil, fmt.Errorf("nil Node in enum type %s", e.Name)
+	if args.contextEntry.Node == nil {
+		return "", "", fmt.Errorf("nil Node in enum type %s", args.contextEntry.Name)
 	}
 
 	// This is the old, buggy behaviour that causes same-named typedefs to shadow one another.
```

Comment:
Can we add a TODO and link to the issue to remove this code once we've transitioned consumers?